File: src/sim/linalg.ts
/**
 * Small dense linear algebra for the simulator.
 *
 * These routines operate on tiny matrices (reduced density matrices are 2x2 or
 * 4x4; the MDS Gram matrix is N x N for N factors, typically <= ~24), so a
 * simple cyclic Jacobi eigensolver is more than adequate and avoids extra deps.
 */

export interface EigenResult {
  /** Eigenvalues, ascending. */
  values: number[]
  /** Eigenvectors as columns: vectors[i][k] is component i of eigenvector k. */
  vectors: number[][]
}

/**
 * Cyclic Jacobi eigensolver for a real symmetric matrix.
 * Returns eigenvalues in ascending order with matching eigenvectors.
 */
export function jacobiEigenSymmetric(input: number[][]): EigenResult {
  const n = input.length
  // Work on a copy.
  const a = input.map((row) => row.slice())
  // Eigenvector accumulator, initialized to identity.
  const v: number[][] = Array.from({ length: n }, (_, i) =>
    Array.from({ length: n }, (_, j) => (i === j ? 1 : 0)),
  )

  const maxSweeps = 100
  for (let sweep = 0; sweep < maxSweeps; sweep++) {
    // Sum of squares of off-diagonal elements.
    let off = 0
    for (let p = 0; p < n; p++) {
      for (let q = p + 1; q < n; q++) off += a[p][q] * a[p][q]
    }
    if (off < 1e-28) break

    for (let p = 0; p < n; p++) {
      for (let q = p + 1; q < n; q++) {
        const apq = a[p][q]
        if (Math.abs(apq) < 1e-300) continue
        const app = a[p][p]
        const aqq = a[q][q]
        const phi = 0.5 * Math.atan2(2 * apq, aqq - app)
        const c = Math.cos(phi)
        const s = Math.sin(phi)

        // Apply rotation J^T A J.
        for (let k = 0; k < n; k++) {
          const akp = a[k][p]
          const akq = a[k][q]
          a[k][p] = c * akp - s * akq
          a[k][q] = s * akp + c * akq
        }
        for (let k = 0; k < n; k++) {
          const apk = a[p][k]
          const aqk = a[q][k]
          a[p][k] = c * apk - s * aqk
          a[q][k] = s * apk + c * aqk
        }
        // Accumulate eigenvectors.
        for (let k = 0; k < n; k++) {
          const vkp = v[k][p]
          const vkq = v[k][q]
          v[k][p] = c * vkp - s * vkq
          v[k][q] = s * vkp + c * vkq
        }
      }
    }
  }

  const values = a.map((row, i) => row[i])
  const order = values
    .map((val, i) => ({ val, i }))
    .sort((x, y) => x.val - y.val)

  const sortedValues = order.map((o) => o.val)
  const sortedVectors: number[][] = Array.from({ length: n }, () =>
    new Array<number>(n).fill(0),
  )
  order.forEach((o, newIdx) => {
    for (let row = 0; row < n; row++) sortedVectors[row][newIdx] = v[row][o.i]
  })

  return { values: sortedValues, vectors: sortedVectors }
}

/**
 * Eigenvalues of a Hermitian matrix H = re + i*im.
 *
 * Uses the standard real embedding M = [[re, -im], [im, re]], a (2n x 2n) real
 * symmetric matrix whose eigenvalues are exactly those of H, each appearing
 * twice. We return the n distinct eigenvalues (ascending).
 */
export function hermitianEigenvalues(re: number[][], im: number[][]): number[] {
  const n = re.length
  const m: number[][] = Array.from({ length: 2 * n }, () =>
    new Array<number>(2 * n).fill(0),
  )
  for (let i = 0; i < n; i++) {
    for (let j = 0; j < n; j++) {
      m[i][j] = re[i][j]
      m[i + n][j + n] = re[i][j]
      m[i][j + n] = -im[i][j]
      m[i + n][j] = im[i][j]
    }
  }
  const { values } = jacobiEigenSymmetric(m)
  // Eigenvalues come in degenerate pairs; take every other (ascending).
  const result: number[] = []
  for (let i = 0; i < 2 * n; i += 2) result.push(values[i])
  return result
}

/**
 * 2D Procrustes alignment: rotate/reflect/scale/translate `x` (N x 2) to best
 * match `y` (N x 2). Used to remove the rotation/reflection gauge freedom of
 * MDS so an evolving 2D embedding does not spin between frames.
 */
export function procrustes2D(x: number[][], y: number[][]): number[][] {
  const n = x.length
  if (n === 0) return x
  const cx = [0, 0]
  const cy = [0, 0]
  for (let i = 0; i < n; i++) {
    cx[0] += x[i][0]
    cx[1] += x[i][1]
    cy[0] += y[i][0]
    cy[1] += y[i][1]
  }
  cx[0] /= n
  cx[1] /= n
  cy[0] /= n
  cy[1] /= n

  const xc = x.map((p) => [p[0] - cx[0], p[1] - cx[1]])
  const yc = y.map((p) => [p[0] - cy[0], p[1] - cy[1]])

  // M = Xc^T Yc (2x2)
  let a = 0
  let b = 0
  let c = 0
  let d = 0
  let normX = 0
  for (let i = 0; i < n; i++) {
    a += xc[i][0] * yc[i][0]
    b += xc[i][0] * yc[i][1]
    c += xc[i][1] * yc[i][0]
    d += xc[i][1] * yc[i][1]
    normX += xc[i][0] * xc[i][0] + xc[i][1] * xc[i][1]
  }

  // Analytic 2x2 SVD of M = U S V^T.
  const e = (a + d) / 2
  const f = (a - d) / 2
  const g = (c + b) / 2
  const h = (c - b) / 2
  const q = Math.hypot(e, h)
  const r = Math.hypot(f, g)
  const sx = q + r
  const sy = q - r
  const a1 = Math.atan2(g, f)
  const a2 = Math.atan2(h, e)
  const theta = (a2 - a1) / 2
  const phi = (a2 + a1) / 2

  // R = U V^T (rotation, possibly with reflection).
  const cu = Math.cos(theta)
  const su = Math.sin(theta)
  const cv = Math.cos(phi)
  const sv = Math.sin(phi)
  const u = [
    [cu, -su],
    [su, cu],
  ]
  const v = [
    [cv, -sv],
    [sv, cv],
  ]
  // R = U * V^T
  const rmat = [
    [u[0][0] * v[0][0] + u[0][1] * v[0][1], u[0][0] * v[1][0] + u[0][1] * v[1][1]],
    [u[1][0] * v[0][0] + u[1][1] * v[0][1], u[1][0] * v[1][0] + u[1][1] * v[1][1]],
  ]
  const scale = normX > 1e-12 ? (sx + sy) / normX : 1

  return xc.map((p) => {
    const rx = p[0] * rmat[0][0] + p[1] * rmat[1][0]
    const ry = p[0] * rmat[0][1] + p[1] * rmat[1][1]
    return [scale * rx + cy[0], scale * ry + cy[1]]
  })
}

/** Shannon/von Neumann entropy (in nats) from a list of probabilities/eigenvalues. */
export function entropyFromEigenvalues(eigs: number[]): number {
  let s = 0
  for (const lambda of eigs) {
    if (lambda > 1e-12) s -= lambda * Math.log(lambda)
  }
  return s
}

File: wasm/mad-dog-sim/src/linalg.rs
//! Small dense linear algebra — port of `src/sim/linalg.ts`.

pub struct EigenResult {
    pub values: Vec<f64>,
    pub vectors: Vec<Vec<f64>>,
}

pub fn jacobi_eigen_symmetric(input: &[Vec<f64>]) -> EigenResult {
    let n = input.len();
    let mut a: Vec<Vec<f64>> = input.to_vec();
    let mut v: Vec<Vec<f64>> = (0..n)
        .map(|i| (0..n).map(|j| if i == j { 1.0 } else { 0.0 }).collect())
        .collect();

    for _ in 0..100 {
        let mut off = 0.0;
        for p in 0..n {
            for q in (p + 1)..n {
                off += a[p][q] * a[p][q];
            }
        }
        if off < 1e-28 {
            break;
        }

        for p in 0..n {
            for q in (p + 1)..n {
                let apq = a[p][q];
                if apq.abs() < 1e-300 {
                    continue;
                }
                let app = a[p][p];
                let aqq = a[q][q];
                let phi = 0.5 * (2.0 * apq).atan2(aqq - app);
                let c = phi.cos();
                let s = phi.sin();

                for k in 0..n {
                    let akp = a[k][p];
                    let akq = a[k][q];
                    a[k][p] = c * akp - s * akq;
                    a[k][q] = s * akp + c * akq;
                }
                for k in 0..n {
                    let apk = a[p][k];
                    let aqk = a[q][k];
                    a[p][k] = c * apk - s * aqk;
                    a[q][k] = s * apk + c * aqk;
                }
                for k in 0..n {
                    let vkp = v[k][p];
                    let vkq = v[k][q];
                    v[k][p] = c * vkp - s * vkq;
                    v[k][q] = s * vkp + c * vkq;
                }
            }
        }
    }

    let values: Vec<f64> = (0..n).map(|i| a[i][i]).collect();
    let mut order: Vec<(f64, usize)> = values.iter().copied().enumerate().map(|(i, val)| (val, i)).collect();
    order.sort_by(|a, b| a.0.partial_cmp(&b.0).unwrap());

    let sorted_values: Vec<f64> = order.iter().map(|o| o.0).collect();
    let mut sorted_vectors: Vec<Vec<f64>> = vec![vec![0.0; n]; n];
    for (new_idx, o) in order.iter().enumerate() {
        for row in 0..n {
            sorted_vectors[row][new_idx] = v[row][o.1];
        }
    }

    EigenResult {
        values: sorted_values,
        vectors: sorted_vectors,
    }
}

/// Full Hermitian eigendecomposition; eigenvectors as interleaved amplitude data.
pub fn hermitian_eigen_decomposition(
    re: &[Vec<f64>],
    im: &[Vec<f64>],
) -> (Vec<f64>, Vec<Vec<f64>>) {
    let n = re.len();
    let mut m = vec![vec![0.0; 2 * n]; 2 * n];
    for i in 0..n {
        for j in 0..n {
            m[i][j] = re[i][j];
            m[i + n][j + n] = re[i][j];
            m[i][j + n] = -im[i][j];
            m[i + n][j] = im[i][j];
        }
    }
    let result = jacobi_eigen_symmetric(&m);
    let mut eigs = Vec::with_capacity(n);
    let mut states = Vec::with_capacity(n);
    for k in 0..n {
        eigs.push(result.values[2 * k]);
        let mut data = vec![0.0; 2 * n];
        for i in 0..n {
            data[2 * i] = result.vectors[i][2 * k];
            data[2 * i + 1] = result.vectors[i + n][2 * k];
        }
        let mut norm = 0.0;
        for i in 0..n {
            norm += data[2 * i] * data[2 * i] + data[2 * i + 1] * data[2 * i + 1];
        }
        norm = norm.sqrt().max(1e-300);
        for v in &mut data {
            *v /= norm;
        }
        states.push(data);
    }
    (eigs, states)
}

pub fn hermitian_eigenvalues(re: &[Vec<f64>], im: &[Vec<f64>]) -> Vec<f64> {
    hermitian_eigen_decomposition(re, im).0
}

pub fn entropy_from_eigenvalues(eigs: &[f64]) -> f64 {
    let mut s = 0.0;
    for &lambda in eigs {
        if lambda > 1e-12 {
            s -= lambda * lambda.ln();
        }
    }
    s
}

/// 2D Procrustes alignment — port of `procrustes2D` in `src/sim/linalg.ts`.
pub fn procrustes_2d(x: &[Vec<f64>], y: &[Vec<f64>]) -> Vec<Vec<f64>> {
    let n = x.len();
    if n == 0 {
        return x.to_vec();
    }
    let mut cx = [0.0; 2];
    let mut cy = [0.0; 2];
    for i in 0..n {
        cx[0] += x[i][0];
        cx[1] += x[i][1];
        cy[0] += y[i][0];
        cy[1] += y[i][1];
    }
    cx[0] /= n as f64;
    cx[1] /= n as f64;
    cy[0] /= n as f64;
    cy[1] /= n as f64;

    let xc: Vec<[f64; 2]> = x
        .iter()
        .map(|p| [p[0] - cx[0], p[1] - cx[1]])
        .collect();
    let yc: Vec<[f64; 2]> = y
        .iter()
        .map(|p| [p[0] - cy[0], p[1] - cy[1]])
        .collect();

    let mut a = 0.0;
    let mut b = 0.0;
    let mut c = 0.0;
    let mut d = 0.0;
    let mut norm_x = 0.0;
    for i in 0..n {
        a += xc[i][0] * yc[i][0];
        b += xc[i][0] * yc[i][1];
        c += xc[i][1] * yc[i][0];
        d += xc[i][1] * yc[i][1];
        norm_x += xc[i][0] * xc[i][0] + xc[i][1] * xc[i][1];
    }

    let e = (a + d) / 2.0;
    let f = (a - d) / 2.0;
    let g = (c + b) / 2.0;
    let h = (c - b) / 2.0;
    let q = (e * e + h * h).sqrt();
    let r = (f * f + g * g).sqrt();
    let sx = q + r;
    let sy = q - r;
    let a1 = g.atan2(f);
    let a2 = h.atan2(e);
    let theta = (a2 - a1) / 2.0;
    let phi = (a2 + a1) / 2.0;

    let cu = theta.cos();
    let su = theta.sin();
    let cv = phi.cos();
    let sv = phi.sin();
    let rmat = [
        [cu * cv + su * sv, cu * -sv + su * cv],
        [su * cv - cu * sv, su * -sv - cu * cv],
    ];
    let scale = if norm_x > 1e-12 { (sx + sy) / norm_x } else { 1.0 };

    xc.iter()
        .map(|p| {
            let rx = p[0] * rmat[0][0] + p[1] * rmat[1][0];
            let ry = p[0] * rmat[0][1] + p[1] * rmat[1][1];
            vec![scale * rx + cy[0], scale * ry + cy[1]]
        })
        .collect()
}

fn mat_mul_3(a: &[[f64; 3]; 3], b: &[[f64; 3]; 3]) -> [[f64; 3]; 3] {
    let mut c = [[0.0; 3]; 3];
    for i in 0..3 {
        for j in 0..3 {
            for k in 0..3 {
                c[i][j] += a[i][k] * b[k][j];
            }
        }
    }
    c
}

fn transpose_3(m: &[[f64; 3]; 3]) -> [[f64; 3]; 3] {
    [
        [m[0][0], m[1][0], m[2][0]],
        [m[0][1], m[1][1], m[2][1]],
        [m[0][2], m[1][2], m[2][2]],
    ]
}

fn det_3(m: &[[f64; 3]; 3]) -> f64 {
    m[0][0] * (m[1][1] * m[2][2] - m[1][2] * m[2][1])
        - m[0][1] * (m[1][0] * m[2][2] - m[1][2] * m[2][0])
        + m[0][2] * (m[1][0] * m[2][1] - m[1][1] * m[2][0])
}

/// 3D Procrustes (Kabsch) alignment — port of `procrustes3D` in `src/sim/linalg.ts`.
pub fn procrustes_3d(x: &[Vec<f64>], y: &[Vec<f64>]) -> Vec<Vec<f64>> {
    let n = x.len();
    if n == 0 {
        return x.to_vec();
    }
    let mut cx = [0.0; 3];
    let mut cy = [0.0; 3];
    for i in 0..n {
        for d in 0..3 {
            cx[d] += x[i].get(d).copied().unwrap_or(0.0);
            cy[d] += y[i].get(d).copied().unwrap_or(0.0);
        }
    }
    for d in 0..3 {
        cx[d] /= n as f64;
        cy[d] /= n as f64;
    }

    let xc: Vec<[f64; 3]> = x
        .iter()
        .map(|p| {
            [
                p.first().copied().unwrap_or(0.0) - cx[0],
                p.get(1).copied().unwrap_or(0.0) - cx[1],
                p.get(2).copied().unwrap_or(0.0) - cx[2],
            ]
        })
        .collect();
    let yc: Vec<[f64; 3]> = y
        .iter()
        .map(|p| {
            [
                p.first().copied().unwrap_or(0.0) - cy[0],
                p.get(1).copied().unwrap_or(0.0) - cy[1],
                p.get(2).copied().unwrap_or(0.0) - cy[2],
            ]
        })
        .collect();

    let mut h = [[0.0; 3]; 3];
    let mut norm_x = 0.0;
    for i in 0..n {
        norm_x += xc[i][0] * xc[i][0] + xc[i][1] * xc[i][1] + xc[i][2] * xc[i][2];
        for j in 0..3 {
            for k in 0..3 {
                h[j][k] += xc[i][j] * yc[i][k];
            }
        }
    }

    let ht = transpose_3(&h);
    let ht_h = mat_mul_3(&ht, &h);
    let ht_h_vec: Vec<Vec<f64>> = ht_h.iter().map(|row| row.to_vec()).collect();
    let eig = jacobi_eigen_symmetric(&ht_h_vec);
    let mut order: Vec<(f64, usize)> = eig
        .values
        .iter()
        .copied()
        .enumerate()
        .map(|(i, val)| (val, i))
        .collect();
    order.sort_by(|a, b| b.0.partial_cmp(&a.0).unwrap());

    let sing: Vec<f64> = order.iter().map(|o| o.0.max(0.0).sqrt()).collect();
    let total_sing: f64 = sing.iter().sum::<f64>().max(1.0);
    let scale = if norm_x > 1e-12 {
        total_sing / norm_x
    } else {
        1.0
    };

    let mut v = [[0.0; 3]; 3];
    for k in 0..3 {
        let col = order[k].1;
        for i in 0..3 {
            v[i][k] = eig.vectors[i][col];
        }
    }

    let mut u = mat_mul_3(&h, &v);
    for k in 0..3 {
        let s = if sing[k] > 1e-12 { sing[k] } else { 1.0 };
        for i in 0..3 {
            u[i][k] /= s;
        }
    }

    let mut rmat = mat_mul_3(&u, &transpose_3(&v));
    if det_3(&rmat) < 0.0 {
        for i in 0..3 {
            u[i][2] *= -1.0;
        }
        rmat = mat_mul_3(&u, &transpose_3(&v));
    }

    xc.iter()
        .map(|p| {
            let rx = p[0] * rmat[0][0] + p[1] * rmat[0][1] + p[2] * rmat[0][2];
            let ry = p[0] * rmat[1][0] + p[1] * rmat[1][1] + p[2] * rmat[1][2];
            let rz = p[0] * rmat[2][0] + p[1] * rmat[2][1] + p[2] * rmat[2][2];
            vec![
                scale * rx + cy[0],
                scale * ry + cy[1],
                scale * rz + cy[2],
            ]
        })
        .collect()
}
